Chore: Enable eslint-plugin-react partial rules

chihsuan opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.
-->

**What this PR does / why we need it**:

This PR enables the following eslint rules in Grafana core (.eslintrc) and fixed the corresponding lint issues:

- react/display-name
- react/no-deprecated
- react/no-find-dom-node


**Which issue(s) this PR fixes**:

Partially for https://github.com/grafana/grafana/issues/29201

**Special notes for your reviewer**:

- I still use inline `// eslint-disable-next-line react/display-name` for some arrow functions since I think rewriting them by function declaration might not be a good idea for coding convention. Correct me, if I'm wrong.

chihsuan committed:
Chore: Enable eslint react/display-name

Enable react/display-name and fixed the corresponding linting issue

part of: #29201

chihsuan committed:
Chore: Enable eslint react/no-deprecated

Enable react/no-deprecated and add the UNSAFE_ prefix for deprected methods

part of: #29201

chihsuan committed:
Chore: Enable eslint react/no-find-dom-node

Enable react/no-find-dom-node rule and use ref instead

part of: #29201

chihsuan committed:
Test: Update TeamGroupSync test snapshot

Since we added the displayName for ToolTip compontent and tag name is changed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=29428) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=29428) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=29428) it.</sub>

chihsuan committed:
Fix: Fixed ClickOutsideWrapper render

The props.children might contains numbers of nodes which make cloneElement failed. Change to simply use a div to wrapper
the children and assign the ref to div for this feature

dprokop commented on packages/grafana-ui/src/components/ColorPicker/SeriesColorPickerPopover.tsx:
```diff
@@ -19,20 +19,22 @@ export const SeriesColorPickerPopover: FunctionComponent<SeriesColorPickerPopove
-              }}
-            />
-          ),
+          tabComponent() {
```
Interesting, is this related to the rules enabled in this PR?

chihsuan replied:
Yes, this is related to the `react/display-name` rule.

Since arrow function is  anonymous, the declared component will not have the displayName. So I changed this to using shorthand method function definition to creates a named function.

dprokop replied:
Perfect, thanks for the explanation:)

dprokop commented on packages/grafana-ui/src/components/OptionsUI/fieldColor.tsx:
```diff
@@ -36,7 +36,9 @@ export const FieldColorEditor: React.FC<FieldConfigEditorProps<FieldColor | unde
       isContinuous: mode.isContinuous,
       isByValue: mode.isByValue,
-      component: () => <FieldColorModeViz mode={mode} theme={theme} />,
+      component: function render() {
+        return <FieldColorModeViz mode={mode} theme={theme} />;
+      },
```
Similar comment to the previous one- how does this approach help the fixes?

chihsuan replied:
This is same like the previous one. I used regular function declaration to have the displayName.
But I just found this could be also written in the shorthand method style. I've changed this to make it consistent.  Thanks!

dprokop reviewed (approved):
@chihsuan thank you so much for this contribution. I went through the code and it seems to be OK! Left two questions I'm interested in getting answer for. Other than that it's awesome!

chihsuan committed:
Style: Use shorthand method definition style for inline component

dprokop commented:
@chihsuan seems like the PR now have a conflict with the master. Can you please sync this with master and I think I can merge it if the build gets green :) Thank you!

chihsuan committed:
Fix: Rebase master and fix linting

Rebase from master branch and fix new displayName linting warning

chihsuan commented:
@dprokop done. please review, thanks!
